File: hw/Vaca/u3/HW30ObserverAndStrategy/HW30StrategyJavaScript/controller/QuickSort.js
export class QuickSort {
    sort(numbers) {
        this.quickSort(numbers, 0, numbers.length - 1);
    }

    quickSort(array, low, high) {
        if (low < high) {
            const pivotIndex = this.partition(array, low, high);
            this.quickSort(array, low, pivotIndex - 1);
            this.quickSort(array, pivotIndex + 1, high);
        }
    }

    partition(array, low, high) {
        const pivot = array[high];
        let i = low - 1;
        for (let j = low; j < high; j++) {
            if (array[j] < pivot) {
                i++;
                [array[i], array[j]] = [array[j], array[i]];
            }
        }
        [array[i + 1], array[high]] = [array[high], array[i + 1]];
        return i + 1;
    }

    getStrategyName() {
        return "Quick Sort";
    }
}

File: hw/Vaca/u3/HW30ObserverAndStrategy/HW30StrategyJava/src/main/java/ec/edu/espe/controller/QuickSort.java

package ec.edu.espe.controller;

/**
 *
 * @author MSI-PULSE
 */
public class QuickSort implements SortingStrategy {
    @Override
    public void sort(int[] numbers) {
        // Implementación del algoritmo de QuickSort
        quickSort(numbers, 0, numbers.length - 1);
    }

    private void quickSort(int[] array, int low, int high) {
        if (low < high) {
            int pivotIndex = partition(array, low, high);
            quickSort(array, low, pivotIndex - 1);
            quickSort(array, pivotIndex + 1, high);
        }
    }

    private int partition(int[] array, int low, int high) {
        int pivot = array[high];
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (array[j] < pivot) {
                i++;
                int temp = array[i];
                array[i] = array[j];
                array[j] = temp;
            }
        }
        int temp = array[i + 1];
        array[i + 1] = array[high];
        array[high] = temp;
        return i + 1;
    }

    @Override
    public String getStrategyName() {
        return "Quick Sort";
    }
}